metron-platform/metron-common/src/main/java/org/apache/metron/common/configuration/StellarStatementReporter.java

```diff
@@ -0,0 +1,166 @@
+        visitEnrichementConfig(child, Type.ENRICHMENT, enrichmentConfig, visitor, errorConsumer);
+        enrichmentConfig = sensorEnrichmentConfig.getThreatIntel();
+        visitEnrichementConfig(child, Type.THREAT_INTEL, enrichmentConfig, visitor, errorConsumer);
+        ThreatTriageConfig threatTriageConfig = sensorEnrichmentConfig.getThreatIntel()
+            .getTriageConfig();
+        visitEnrichmentThreatTriageConfigs(child, threatTriageConfig, visitor, errorConsumer);
+      } catch (Exception e) {
+        List<String> names = Arrays.asList(getName(), ENRICHMENT.toString(), child);
+        errorConsumer.consume(names, e);
+      }
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  private void visitEnrichementConfig(String topicName, Type type,
+      EnrichmentConfig enrichmentConfig, StatementReportVisitor visitor,
+      ConfigReportErrorConsumer errorConsumer) throws Exception {
+
+    Map<String, Object> enrichmentStellarMap = (Map<String, Object>) enrichmentConfig.getFieldMap()
+        .getOrDefault("stellar", new HashMap<>());
+    Map<String, Object> transforms = (Map<String, Object>) enrichmentStellarMap
+        .getOrDefault("config", new HashMap<>());
+    try {
+      for (Map.Entry<String, Object> kv : transforms.entrySet()) {
+        // we can have a group or an entry
+        if (kv.getValue() instanceof Map) {
+          Map<String, String> groupMap = (Map<String, String>) kv.getValue();
+          for (Map.Entry<String, String> groupKv : groupMap.entrySet()) {
+            List<String> names = Arrays
+                .asList(getName(), ENRICHMENT.toString(), topicName, type.toString(), kv.getKey(),
+                    groupKv.getKey());
+            visitor.visit(names, groupKv.getValue());
+          }
+        } else {
+          List<String> names = Arrays
+              .asList(getName(), ENRICHMENT.toString(), topicName, type.toString(), "(default)",
+                  kv.getKey(), kv.getKey());
+          visitor.visit(names, kv.getValue().toString());
+        }
+      }
+    } catch (Exception e) {
+      List<String> names = Arrays
+          .asList(getName(), ENRICHMENT.toString(), topicName, type.toString());
+      errorConsumer.consume(names, e);
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  private void visitEnrichmentThreatTriageConfigs(String topicName,
+      ThreatTriageConfig threatTriageConfig, StatementReportVisitor visitor,
+      ConfigReportErrorConsumer errorConsumer) throws Exception {
+    try {
+      List<RiskLevelRule> riskLevelRules = threatTriageConfig.getRiskLevelRules();
+      riskLevelRules.forEach((r) -> {
+        String name = r.getName();
+        if (org.apache.commons.lang.StringUtils.isEmpty(name)) {
+          name = "(default)";
+        }
+        List<String> names = Arrays
+            .asList(getName(), ENRICHMENT.toString(), topicName, "THREAT_TRIAGE", name);
+        visitor.visit(names, r.getRule());
```

Comment:
I like what you are doing with this PR.

There is just one downside that bugs me. How do we ensure that as our configuration evolves over time that we also update this to ensure it gets validated?  

We have Stellar expressions as configurations all over the place; parsing, enrichment, triage, indexing, and profiler.  I feel like any/all of these will evolve and change over time.  I would definitely forget to come here and update this to make sure it gets validated.

As a motivating example, right here you are only validating the `rule` field.  But triage also has the `reason` field that is a Stellar expression.  That would need to be validated also.  That was something we added later and likely the scenario where I would forget to add that expression to your validation logic.

Reply:
And let me follow-on with one proposal that might help address this.  There are probably other (better?) ways to solve this, but here is one approach to chew on.

All of our configuration gets deserialized from JSON into POJOs; EnrichmentConfig, ProfilerConfig, etc.  What if we had an annotation that we marked the fields that are required to be valid Stellar expressions?  Based on the annotation we can then validate each of those fields.

Since the annotation is directly in the configuration classes, it is less likely I'm going to forget that annotation.  And it is also remains decoupled, which is a good benefit of your current approach.

Reply:
I need to think about this, but I will add the reason check

Reply:
This also makes it simple to validate the configuration, no matter where it rests.  

To validate the configuration in Zk, you just read all the configs in from Zk, deserialize into POJOs, iterate all the fields looking for the annotation, if you find the annotation, then run your validation logic on the expression.

To validate the configuration on the file system, you just read it all in from the FS, then repeat the same steps I described in the Zk scenario.

Reply:
So - if I'm understanding you correctly, as it applies to what I am doing ->
when visiting configs, instead of explicitly validating fields, we would want to 'visit' all the members per pojo by attribute.  We would only need to worry about tracking the pojos but not the fields ( we need to keep the context ).  Am I understanding you correctly?

Reply:
When I say track pojo's I mean the reporter would still 'know' about the Configuration object, but would discover what fields where stellar.  Know the object but not the fields.

Reply:
you would either have to 'know' about fields that were other nested types or do a recursive desert through the pojo though

Reply:
> you would either have to 'know' about fields that were other nested types or do a recursive desert through the pojo though

Apache Commons has a `FieldUtils.getFieldsListWithAnnotation(...)`.  Once you get a 'configuration' class, then you just use that to get all the fields that have been marked with the annotation.

Reply:
> We would only need to worry about tracking the pojos but not the fields ( we need to keep the context ). Am I understanding you correctly?

Yes, exactly.  You can use the functionality in `ConfigurationUtils` and just do something like....
``` 
config = getProfilerConfig(...)
validateStellar(config)

config = getSensorConfig(...)
validateStellar(config)

config = getEnrichmentConfig(..)
validateStellar(config)

...
```

Or something to that effect.

Reply:
does profiler have stellar?  ( sorry to have to ask ).

Reply:
True, but FieldUtils does a class, I don't think it recurses.
So we would still need to do some work

Reply:
> does profiler have stellar? ( sorry to have to ask ).

No problem. :)  Yes, it is mostly Stellar.  The docs are pretty good on that one, but I can outline specifically which ones if that helps.

Reply:
> True, but FieldUtils does a class, I don't think it recurses.  So we would still need to do some work

I think the logic would be something like...
* get the configuration object
* find what class the configuration object is
* use `FieldUtils..getFieldsListWithAnnotation` to get all the fields that are Stellar
* then for those fields that are Stellar, get the value from the configuration object 
* that value is the stellar expression.  validate it.

Reply:
right, for SensorEnrichmentConfig, I would have to know about the 'inner' config objects and do them to, just as I do now.

Reply:
Gotcha.  Understand your point now.  Seems like we should be able to handle that in a generic way for any configuration object.

Reply:
@nickwallen, checking the reason field....  Because the rules are checked on deserialize, if they are wrong, the whole config fails ( the entire Enrichment ).  You *could* say that the checks that I do have are more than I need, and that rather than add reason, I should just get rid of the rules stuff all together.  What do you think?

Reply:
@nickwallen you know what would work better maybe?  if we could attribute the JSON.

Reply:
> because the rules are checked on deserialize, if they are wrong, the whole config fails ( the entire Enrichment )

  Only the syntax is checked on deserialization now.  If I add a field that we don't expect, we will catch that.  But if I use invalid Stellar; example use a variable named 'default', we will NOT catch that.  That's what your work comes in.  

I am not sure that I am understanding your suggestion though.  Please correct me if this is a miss.

Reply:
> you know what would work better maybe? if we could attribute the JSON.

Maybe not?  Users/admins of the system create the JSON.  I don't want them to annotate which field is Stellar. I want the developers, the creators of the extension points, to define which fields are Stellar.

Reply:
Nick, when the ThreatTriage is loaded from json, the call to setRiskLevelRules() gets called by jackson, which validates the stellar, which will throw the exception at load time, before I can do the check myself.

So I have code to verify the rules, but the fact is if they are bad stellar, they will throw on load before that code will execute.

Reply:
Sure, I see in that case validation does exist.  And this highlights an alternative approach.  Add custom code like that to each of the POJOs to validate any Stellar fields.  Personally, I like the annotation based approach.  

But whichever way we do it, we'd probably want to see it done the same across the board.  There are many other places where no validation of Stellar fields happen, right?

Reply:
I think we eventually want to get to that point, with a consistent, comprehensive mechanism to validate stellar statements ( short of compiling ;) ).

Reply:
I *think* I know how to do it and not have it be terrible.  The question is if it should be in this PR or after

Reply:
I think I can do it such that the stellar-common holds the code for doing the traversals as well as the interfaces and callbacks, and the *reporter* just provides the configs.

I'm thinking it through, but now that I'm thinking of it I will be miserable until I do it so I'll do it.

Reply:
I think you are really going to like this @nickwallen

Reply:
but if you complain about scope or the review size @nickwallen I will be very cross